Feature/36 create chapter 3 (#37)
cloudogu/dogu-development-docs: docs/important/relevant_functionalities_de.md

    - Mit dem Schalter `--timeout <Sekunden>` kann man so auf Dogus warten, von denen das eigene Dogu abhängt.
    - Ein Exit-Code ungleich Null (0) deutet darauf hin, dass entweder das Dogu nicht rechtzeitig seinen Startvorgang beendet hat oder die Registry-Kommunikation fehlschlug.

## Aufbau und Best Practices von `startup.sh` 

Ein einfaches Dogu benötigt eigentlich nur eine `dogu.json` und ein Container-Image. Doch was geschieht, wenn sich die Wirklichkeit im Dauerbetrieb von den Annahmen während der Dogu-Entwicklung unterscheiden? 

Bei Cloudogu haben wir schnell festgestellt, dass es im Betrieb geschickter ist, auf solche Änderungen eingehen zu können. Das Spektrum ist hierbei breit. Dabei kann es sich um Wartezeiten gegenüber anderen Dogus handeln, oder um die Änderung von ursprünglich fixen Bezeichnern bis hin zu Störungen in der Container-Engine. 

Wiederkehrende Aufgaben beim Container-Start befinden sich im Abschnitt "Typische Dogu-Features" in den Abschnitten:
- [Änderbarkeit der Admingruppe](#änderbarkeit-der-admin-gruppe)
- [Änderbarkeit der FQDN](#änderbarkeit-der-fqdn)
- [Loglevel mappen und ändern](#loglevel-mappen-und-ändern)

> Um dynamisch auf diese Gegebenheiten zu reagieren hat es sich bei Cloudogu eingebürgert, nicht den Hauptprozess im Container direkt zu starten. Stattdessen werden erst Fehlerfälle in einem Startskript abgeprüft, ggf. neue Konfigurationswerte einarbeitet, um dann erst den Hauptprozess zu starten, z. B. so:

Reviewer: ```suggestion
Um dynamisch auf diese Gegebenheiten zu reagieren hat es sich bei Cloudogu eingebürgert, nicht den Hauptprozess im Container direkt zu starten. Stattdessen werden erst Fehlerfälle in einem Startskript abgeprüft, ggf. neue Konfigurationswerte eingearbeitet, um dann erst den Hauptprozess zu starten, z. B. so:
```

> Um dynamisch auf diese Gegebenheiten zu reagieren hat es sich bei Cloudogu eingebürgert, nicht den Hauptprozess im Container direkt zu starten. Stattdessen werden erst Fehlerfälle in einem Startskript abgeprüft, ggf. neue Konfigurationswerte einarbeitet, um dann erst den Hauptprozess zu starten, z. B. so:

Author: done

1. Container startet `startup.sh`
2. `startup.sh`
   1. setzt eigene Optionen
   2. ruft eigene Kommandos auf, um bestimmte Funktionalitäten umzusetzen, um z. B.
      - auf Dogu-Abhängigkeiten zu warten
      - einmalige Installationsprozesse durchzuführen 
      - ein temporäres Admin-Konto zu generieren

>       - das aktuelle Log-Level umsetzen

Reviewer: ```suggestion
      - das aktuelle Log-Level umzusetzen
```

>       - Registrieren der CES-SSL-Zertifikate in der Software/Zielsystem

Author: done

      - Vorarbeiten mit API-Zugriffen, um die Software vorzubereiten
        - Replikation von LDAP-Gruppen / Rechten
      - Einstellen der Software (Log-Level, sonstige Konfiguration, z. B. durch `doguctl template`)

>       - Auf Systemänderungen der Registry reagieren: z. B. [Admin-Gruppe](#änderbarkeit-der-admin-gruppe) hat sich geändert, das im Dogu weitere Maßnahmen benötigt

Reviewer: ```suggestion
      - Auf Systemänderungen der Registry reagieren: z. B. [Admin-Gruppe](#änderbarkeit-der-admin-gruppe) hat sich geändert, was im Dogu weitere Maßnahmen benötigt
```

Author: done

   3. den Dogu-State auf `ready` setzen (geeigneter [HealthCheck](../core/compendium_de.md#healthchecks) vorausgesetzt)
   4. startet den Hauptprozess

Dieser Abschnitt geht daher auf Erkenntnisse und _Best Practices_ ein, die sich auf solche Startskripte beziehen: Die `startup.sh`. 

Übrigens: Fleißige Entwickler:innen können Inspiration in den Cloudogu-eigenen Startskripten [z. B. im Redmine-Dogu](https://github.com/cloudogu/redmine/blob/develop/resources/startup.sh) sammeln.

### Skriptinterpreter

Um ein Skript in einem Dogu ausführen zu können, muss ein Skriptinterpreter im Container-Image existieren. Das kann ein offizielles Paket sein (wie `bash`), es spricht aber nichts dagegen seinen eigenen Skriptinterpreter zu verwenden. Da Bash-Skripte weit verbreitet sind, wird hier Bash-Syntax verwendet.

### Fehlerbehandlung

> Je nach Dogu können Fehler an unterschiedlichen Stellen entstehen und (dummerweise) auch wieder verschluckt werden. Es ist eine gute Praxis, Fehler hart aufschlagen zu lassen, um die Fehlerursache schneller zu identifizieren.

Reviewer: ```suggestion
Je nach Dogu können Fehler an unterschiedlichen Stellen entstehen und (dummerweise) auch wieder verschluckt werden. Es ist eine gute Praxis, das Skript bei Fehlern mit einem entsprechenden exit-Code abzubrechen, um die Fehlerursache schneller zu identifizieren.
```

Author: done

> Je nach Dogu können Fehler an unterschiedlichen Stellen entstehen und (dummerweise) auch wieder verschluckt werden. Es ist eine gute Praxis, Fehler hart aufschlagen zu lassen, um die Fehlerursache schneller zu identifizieren.

Author: Please read the other comments as well to avoid duplicates

Hierzu werden als allererstes die folgenden Optionen gesetzt

```bash
#!/bin/bash

> set -o errexit # beende das gesamte Skript (und damit auch den Container) bei einem nicht abgefangenem Fehler 

Author: done

set -o nounset # finde nicht initialisierte Variablen
set -o pipefail # verschlucke keine Fehler bei Pipe-Verwendung
```

Wie ein Fehler abgefangen wird, ist ein Implementierungsdetail. Manchmal ist es günstig, einen Fehlerfall zu testen und eine eigene Meldung abzusetzen, manchmal reicht der Fehler selbst aus.

Wenn man bereits Fehler erwartet, dann kann dieses Konstrukt hilfreich sein:

```bash
yourCommandExitCode=0
your-command || yourCommandExitCode=$?
if [[ ${yourCommandExitCode} -ne 0 ]]; then

>   echo "ERROR: Oh no. I found something critical during QuirkyFunction.";

Author: done

>   doguctl state "ErrorQuirkyFunction"

Author: done

### Teile und herrsche mit Bash

Je länger man eine Anwendung entwickelt, desto mehr Funktionalitäten gelangen in die Anwendung. Genauso kann es sich auch mit der `startup.sh` verhalten. Von einer leichteren Testbarkeit (z. B. mit [BATS](https://github.com/bats-core/bats-core)) abgesehen, sind kleine Ausführungsbestandteile leichter verständlich.

Es ist daher eine gute Idee, Dogu-Funktionalitäten in der `startup.sh` genauso unter Gesichtspunkten der Testbarkeit, Lesbarkeit oder Refaktorisierbarkeit zu gestalten.

#### Shell functions

Hierbei helfen Bash-Funktionen:

```bash
function setDoguLogLevel() {
  echo "Mapping dogu specific log level..."
  currentLogLevel=$(doguctl config --default "WARN" "logging/root")
  
  # bilde hier das Loglevel auf die Logkonfiguration des Dogus ab 
}
```

Wiederkehrende Funktionen können abstrahiert und parametrisiert werden:

```bash
function deleteSetupArtefact() {
  local artefactFile="${1}"
  rm -f "${artefactFile}"

> }

Author: I am game.

> }

Author: done

function cleanUpSetup() {
  deleteSetupArtefact "/tmp/tracker-extract-3-files.127"
  deleteSetupArtefact "/usr/share/lib/intermediateFile.bin"

> }

Author: done

```

#### Nachvollziehbarkeit während der Startphase

Mit steigender Komplexität ist es eine Idee wert, relevante Schritte mit einem `echo` zu versehen, um im Fehlerfall eine Suche nach dem Fehler in den Logausgaben zu beschleunigen.

Zeitgleich gibt es mit dem [Dogu-State-HealthCheck](../core/compendium_de.md#healthchecks) die Möglichkeit, während des Dogu-Starts eigene Ausführungsphasen zu markieren. Zum Beispiel `doguctl state 'aSpecialInstallPhase'` am Anfang oder `doguctl state 'ready'` am Ende des Skripts.

> Der Zustand `ready` ist für die Interpretation des Health-Checks zwingend notwendig.

Reviewer: ```suggestion
Ein HealthCheck vom Typ "state" wird ausschließlich dann erfolgreich sein, wenn der state den String `ready` enthält.
```

Author: done

#### Auslagerung von Code

> Eine weitere Möglichkeit ist die Auslagerung von Funktionen in andere Skriptdateien, die dann mittels `source` in das eigentliche Startskript inkludiert werden. Es sollte allerdings beachtet werden, dass auch ein solches `source`n Fehlschlagen kann und behandelt werden sollte.

Reviewer: ```suggestion
Eine weitere Möglichkeit ist die Auslagerung von Funktionen in andere Skriptdateien, die dann mittels `source` in das eigentliche Startskript inkludiert werden. Es sollte allerdings beachtet werden, dass auch ein solches `source`n fehlschlagen kann und behandelt werden sollte.
```

Author: done

```bash
sourcingExitCode=0
source /resources/util.sh || sourcingExitCode=$?
if [[ ${sourcingExitCode} -ne 0 ]]; then
  echo "ERROR: An error occurred while sourcing /resources/util.sh: Exit code ${sourcingExitCode}"
  doguctl state "ErrorSourcingUtilSh"
  sleep 300
  exit 1
fi
```

### Qualitätssicherung von Startskripten

Bei komplexeren Aufgaben kann es sein, dass ein Startskript wie die `startup.sh` an Komplexität gewinnt. Wie jedes andere Programm sollte es getestet und auf Robustheit analysiert werden. Hierbei gibt es mehrere Möglichkeiten. [BATS](https://github.com/bats-core/bats-core) kann im Unit-Test von Shell-Funktionen unterstützen.

> Analysewerkzeuge wie [Shellcheck](https://www.shellcheck.net/) können Fehler in der Benutzung von Shell-Aufrufen aufdecken.

Reviewer: Maybe we should add that passing a shellcheck test is an absolute must-have for dogu scripts, as we will also test this in the Jenkins pipelines.

Author: We can only suggest so much. How an external dogu developer develops her CI is completely outside our control

Reviewer: I think shellcheck is one of those things we will enforce later, as it is not Dogu-specific and can always be checked on any external dogu code

Reviewer: all right, this can be added later

### Die Nutzung von `doguctl`

> Der Abschnitt [über Registry-Zugriff](#registry-zugriff-vom-dogu-heraus) hat das Thema `doguctl` bereits angeschnitten. `doguctl` ist ein Kommandozeilenwerkzeug, das wiederkehrende Interaktionen mit dem Cloudogu EcoSystem bündelt und vereinfacht. Dieser Abschnitt beschreibt mögliche Aufrufe.

Reviewer: ```suggestion
Der Abschnitt [über Registry-Zugriff](#registry-zugriff-vom-dogu-heraus) hat das Thema `doguctl` bereits angeschnitten. `doguctl` ist ein Kommandozeilenwerkzeug, das wiederkehrende Interaktionen mit der etcd-Registry des Cloudogu EcoSystem bündelt und vereinfacht. Dieser Abschnitt beschreibt mögliche Aufrufe.
```

Author: done

> Mit `--help` gibt jedes Unterkommando von `doguctl` gibt eine Hilfeseite aus.  

Author: done

#### doguctl config

Dieser Aufruf liest und schreibt Konfigurationswerte.

```bash
# liest Konfigurationwert aus dem Schlüssel /config/<dogu>/my_key
doguctl config my_key
my old value

# liest globalen Konfigurationwert aus /config/_global/fqdn
doguctl config -g fqdn
your-ecosystem.example.com

> # liest Konfigurationwert aus /config/<dogu>/my_key, gibt einen Defaultwert zurück, wenn dieser nicht gesetzt wurde

Author: done

doguctl config --default NOVALUE my_key2
NOVALUE

# liest verschlüsselten DB-Namen /config/<dogu>/sa-postgresql/db_name
doguctl config -e sa-postgresql/db_name
your-dogu-database-1234

# schreibt Konfigurationswert in /config/<dogu>/my_key hinein
doguctl config my_key 'my new value'

> # schreibt verschlüsselten Geheimnis /config/<dogu>/geheim/credential

Reviewer: ```suggestion
# schreibt verschlüsseltes Geheimnis /config/<dogu>/geheim/credential
```

Author: done

```

#### doguctl validate

Dieser Aufruf validiert Konfigurationswerte der Registry. Das Kommando `validate` liefert einen Exit Code == 0, wenn keines der überprüften Dogu-Konfigurationswerte einen Validierungsfehler aufweist. Ansonsten liefert es einen Exit Code == 1.

```bash
doguctl validate logging/root # validiert einzelnen Wert, Gutfall: Wert=WARN aus ERROR,WARN,INFO,DEBUG
Key logging/root: OK
echo $?
0

doguctl validate logging/root #  validiert einzelnen Wert, Fehlerfall: Wert=foo aus ERROR,WARN,INFO,DEBUG
Key logging/root: INVALID
at least one key was invalid
echo $?
1

doguctl validate --all # validiert alle Werte mit Ausgabe
...
Key container_config/memory_limit: OK
Key container_config/swap_limit: OK
Key container_config/java_max_ram_percentage: OK
Key container_config/java_min_ram_percentage: OK
Key logging/root: INVALID
Key caching/workdir/size: OK (No Validator)
at least one key was invalid
echo $?
1

doguctl validate --silent # validiert Werte ohne Ausgabe (kombinierbar mit --all), Gutfall
echo $?
0

doguctl validate --silent # validiert Werte ohne Ausgabe (kombinierbar mit --all), Fehlerfall
at least one key was invalid
echo $?
1
```

#### doguctl healthy

> Dieser Aufruf prüft, ob ein gegebenes Dogu betriebsbereit (healthy) ist. Das Kommando `healthy` liefert einen Exit Code == 0, wenn das betrachtete Dogu healthy ist. Ansonsten liefert es einen Exit Code == 1.

Reviewer: ```suggestion
Dieser Aufruf prüft, ob ein gegebenes Dogu betriebsbereit (healthy) ist. Das Kommando `healthy` liefert einen Exit Code == 0, wenn alle HealthChecks des betrachteten Dogus ein positives Ergebnis zurückgeben. Ansonsten liefert es einen Exit Code == 1.
```

Author: done

$ echo $?
0

$ doguctl validate --silent # validiert Werte ohne Ausgabe (kombinierbar mit --all), Fehlerfall
at least one key was invalid
echo $?
1
```

#### doguctl healthy

Dieser Aufruf prüft, ob ein gegebenes Dogu betriebsbereit (healthy) ist. Das Kommando `healthy` liefert einen Exit Code == 0, wenn das betrachtete Dogu healthy ist. Ansonsten liefert es einen Exit Code == 1.

```bash
if ! doguctl healthy --wait --timeout 120 postgresql; then
  echo "timeout reached by waiting of ldap to get healthy"
  exit 1
fi
```

#### doguctl state

Dieser Aufruft liest und schreibt Dogu-Zustandswerte. Er wird in Kombination mit einem [HealthCheck](../core/compendium_de.md#healthchecks) verwendet.

```bash
$ doguctl state "installing" # schreibt den Wert in den State
# Erledige Installation

$ doguctl state # Liest den Wert aus dem State
installing

$ doguctl state "ready" # Setzt den State auf den Standard-Healthy-Wert
# Starte nun die Anwendung
```

#### doguctl random

Dieser Aufruf erzeugt Zufallsstrings, geeignet um Passwörter oder sonstige Zugangsdaten zu erzeugen.

```bash
$ doguctl random # erzeugt Zufallsstring mit der Länge 16
9HoF4nYmlLYtf6Ju

> $ doguctl random  -l 5 # erzeugt Zufallsstring mit gegebener Länge

Reviewer: ```suggestion
$ doguctl random  -l 5 # erzeugt Zufallsstring mit Länge 5
```

Author: done

Reviewer: No its not :D

W6Wmj

> ```

Reviewer 2: is this cryptographically secure?

Author: From a theoretical stance: not yet. For an cryptgraphically secure value the go library `crypto` must be used. 

From a practical stance: Safe enough given the unpredictability of a container runtime latency.

Der Aufruf nimmt ein bis zwei Parameter entgegen:
```bash
$ doguctl template <Template-Datei> [Ausgabedatei]
```

Wird das Kommando `template` ohne Ausgabedatei verwendet, dann wird die gerenderte Ausgabe auf `stdout` ausgegeben.

Unterstützte Template-Parameter:
- Umgebungsvariablen
  - `.Env.Get <Umgebungsvariable>` - verwendet vorher exportierte Umgebungsvariable "ADMIN_USERNAME"
- Dogu-Konfiguration
  - `.Config.Get <Dogu-Konfigurationsschlüssel>` - verwendet unverschlüsselten Dogu-Konfigurationswert unter `/config/<dogu>/<schlüssel>` 
  - `.Config.GetOrDefault <Dogu-Konfigurationsschlüssel> <Fehlwert>` - verwendet unverschlüsselten Dogu-Konfigurationswert oder den bereitgestellten Fehlwert
  - `.Config.GetAndDecrypt <verschlüsselter Dogu-Konfigurationsschlüssel>` - entschlüsselt den Dogu-Konfigurationswert und verwendet ihn
  - `.Config.Exists <Dogu-Konfigurationsschlüssel>` - gibt einen `bool`-Wert zurück, ob ein Dogu-Konfigurationswert unter `/config/<dogu>/<schlüssel>` existiert
- Globale Konfiguration
  - `.GlobalConfig.Get <globaler Konfigurationsschlüssel>` - verwendet globalen Konfigurationswert `/config/_global/<schlüssel>`

>   - `.GlobalConfig.GetOrDefault <globaler Konfigurationsschlüssel> <Fehlwert>` - verwendet unverschlüsselten globalen Konfigurationswert oder den bereitgestellten Fehlwert

Reviewer: ```suggestion
  - `.GlobalConfig.GetOrDefault <globaler Konfigurationsschlüssel> <Defaultwert>` - verwendet unverschlüsselten globalen Konfigurationswert oder den bereitgestellten Defaultwert
```

Author: done (in the whole file)

  - `.GlobalConfig.Exists <Dogu-Konfigurationsschlüssel>` - gibt einen `bool`-Wert zurück, ob ein globaler Konfigurationswert unter `/config/_global/<schlüssel>` existiert
- Dogu-Abfragen
  - `.Dogus.IsEnabled <Dogu-Name>` - gibt einen `bool`-Wert zurück, ob ein Dogu installiert ist
  - `.Dogus.Get <Dogu-Name>` - gibt ein [Dogu](../core/compendium_de.md#type-dogu)-Struct zurück

>   - `.Dogus.GetAll` - gibt ein Slice aller [Dogus](../core/compendium_de.md#type-dogu) zurück

Reviewer 2: I don't know if everybody knows the meaning of slice if they didn't get in touch with Go before. Maybe add a hint for that?
```suggestion
  - `.Dogus.GetAll` - gibt ein Slice (eine Art Array in Go) aller [Dogus](../core/compendium_de.md#type-dogu) zurück
```

Author: Good point, but I do like to call a slice an array. Rather I'd should go about that the templating mechanism is Golang templating so then we can expect people to handle terms like this.

Author: Edit: In the very first paragraph we already tell people that this is a Golang template. I think this can stay this way.

Beispiel einer Template-Datei:

```gotemplate
# my-pseudo-config.conf
[Global]
admin.username = {{ .Env.Get "ADMIN_USERNAME" }}
funny.name = {{ .Config.Get "something_funny" }}
log.level = {{ .Config.GetOrDefault "log_level" "WARN" }}
url.jdbc = jdbc:postgresql://postgresql:5432/{{ .Config.GetAndDecrypt "sa-postgresql/database" }}
url.fqdn = https://{{ .GlobalConfig.Get "fqdn" }}/my-dogu 

{{ if .Config.Exists "notification" }}
notification = mail
{{ end }}

{{ if .Dogus.IsEnabled "redmine" }}
ticketsystem.url = https://{{ .GlobalConfig.Get "fqdn" }}/redmine
{{ end }}
...
```

#### doguctl wait-for-tcp

Mit diesem Aufruf wartet `doguctl` bis ein gegebener TCP-Port offen ist oder ein Timeout (in Sekunden) eintritt. Ein Exit Code != 0 signalisiert einen Fehler.

```bash
if ! doguctl wait-for-tcp --timeout 120 8080; then

>   echo "Received timeout during wait for port 8080. Exiting."

Reviewer: ```suggestion
  echo "Received timeout after waiting 120 seconds for port 8080. Exiting."
```

Author: done

  exit 1
fi
```

#### doguctl wait-for-http

> Mit diesem Aufruf wartet `doguctl` bis eine gegebene HTTP-URL bereit ist oder ein Timeout (in Sekunden) eintritt. Ein Exit Code != 0 signalisiert einen Fehler.

Reviewer 2: Maybe we should be open here for HTTPs?
```suggestion
Mit diesem Aufruf wartet `doguctl` bis eine gegebene HTTP(s)-URL bereit ist oder ein Timeout (in Sekunden) eintritt. Ein Exit Code != 0 signalisiert einen Fehler.
```

Author: done

```bash
if ! doguctl wait-for-tcp --host postgresql --timeout 120 5432; then

>   echo "Received timeout during wait for port 5432 of host postgresql. Exiting."

Reviewer: ```suggestion
  echo "Received timeout after waiting 120 seconds for port 5432 of host postgresql. Exiting."
```

Author: done

  exit 1
fi
```

## Service Accounts

Service Accounts bilden einen Mechanismus für Zugangskonten ab, die Dogus zur Absicherung oder Speicherung ihrer Daten benötigen, diese Funktionalität aber nicht selbst bereitstellen möchten. Dogus können sich als Produzent und/oder als Konsument von Service Accounts darstellen. Bei beiden handelt es sich um ein jeweils optionales Feature eines Dogus. 

> In der Regel diktiert die Natur der Applikation im Dogu selbst, ob bzw. welches von beidem am sinnvollsten ist. So wäre ein Datenbankmanagement-Dogu wie PostgreSQL nicht hilfreich, das anderen Dogus keine Zugänge zu seiner Datenbasis anbietet. Andererseits wäre es für ein Dogu wie Redmine fatal, kein Dogu zu haben, in das es per SQL seine Daten ablegen kann.

Reviewer 2: ```suggestion
In der Regel diktiert die Natur der Applikation im Dogu selbst, ob bzw. welches von beiden am sinnvollsten ist. So wäre ein Datenbankmanagement-Dogu wie PostgreSQL nicht hilfreich, das anderen Dogus keine Zugänge zu seiner Datenbasis anbietet. Andererseits wäre es für ein Dogu wie Redmine fatal, kein Dogu zu haben, in das es per SQL seine Daten ablegen kann.
```

Author: done

> Dogus, die anderen Dogus gegenüber Zugangskonten bereitstellen können, nennt man auch Service-Account-Produzent. Dogus, die ein Zugangskonto bei einem anderen Dogu benötigen, nennt man Service-Account-Konsument.

Reviewer: ```suggestion
Dogus wie PostgreSQL, die anderen Dogus gegenüber Zugangskonten bereitstellen können, nennt man auch Service-Account-Produzent. Dogus wie Redmine, die ein Zugangskonto bei einem anderen Dogu benötigen, nennt man Service-Account-Konsument.
```

Author: done

> Dogus, die anderen Dogus gegenüber Zugangskonten bereitstellen können, nennt man auch Service-Account-Produzent. Dogus, die ein Zugangskonto bei einem anderen Dogu benötigen, nennt man Service-Account-Konsument.

Reviewer 2: ```suggestion
Dogus, die anderen Dogus gegenüber Zugangskonten bereitstellen können, nennt man auch Service-Account-Produzenten. Dogus, die ein Zugangskonto bei einem anderen Dogu benötigen, nennt man Service-Account-Konsumenten.
```

Author: done

### Service Accounts produzieren 

> Service Accounts werden nicht von einem Konsumer-Dogu selbst angefragt, weil ein Dogu sich nicht um diese Form der Orchestrierung kümmern sollte. Dies ist stattdessen die Aufgabe von einem Client wie `cesapp` oder `k8s-dogu-operator` während der Installation des Service-Account-Konsument-Dogus. Das `ExposedCommand` hierfür lautet `service-account-create`. 

Reviewer 2: Kurz zuvor eingeführte Begriffe werden hier nicht genutzt.
Sind Service-Account eigentlich nur für Dogus oder später auch Operators o.ä.? In Kubernetes gibt es ja automatische Service-Bindings z.B. für Datenbanken für sowas.
```suggestion
Service Accounts werden nicht von einem Service-Account-Konsumenten selbst angefragt, weil ein Dogu sich nicht um diese Form der Orchestrierung kümmern sollte. Dies ist stattdessen die Aufgabe von einem Client wie `cesapp` oder `k8s-dogu-operator` während der Installation des Service-Account-Konsumenten. Der `ExposedCommand` hierfür lautet `service-account-create`. 
```

Author: done

Ähnlich hierzu sieht die Löschung von Service Accounts aus, denn ebenfalls kümmert sich ein Client wie `cesapp` oder `k8s-dogu-operator` während der Deinstallation eines Dogus darum, dass dessen angeforderte Service Accounts wieder gelöscht werden. Das `ExposedCommand` hierfür lautet `service-account-remove`.

> Damit ein Dogu als Service-Account-Produzent auftreten kann, muss es in seiner `dogu.json` zwei ausführbare Dateien exponieren:

Reviewer 2: ```suggestion
Damit ein Dogu als Service-Account-Produzent auftreten kann, muss es in seiner `dogu.json` zwei ausführbare Dateien bereitstellen:
```

Author: done

```json
{
  ...,
  "ExposedCommands": [
    {
      "Name": "service-account-create",
      "Description": "Creates a new service account",
      "Command": "/create-sa.sh"
    },
    {
      "Name": "service-account-remove",
      "Description": "Removes a service account",
      "Command": "/remove-sa.sh"
    }
  ],
  ...,
}
```

> Die Begriffe `service-account-create` und `service-account-remove` sind geschützte Eigenschaften und werden von Clients, die die `dogu.json` interpretieren als Befehle zum Erstellen und Löschen von Service Accounts benutzt.

Reviewer: ```suggestion
Die Begriffe `service-account-create` und `service-account-remove` sind geschützte Eigenschaften und werden von Clients, die die `dogu.json` interpretieren, als Befehle zum Erstellen und Löschen von Service Accounts benutzt.
```

Author: done

In den beiden folgenden Abschnitten wird auf die Erzeugung und Löschung von Service Accounts eingegangen. Weitere Information über ExposedCommands befinden sich im [ExposedCommands](../core/compendium_de.md#exposedcommands)-Abschnitt des Kompendiums.

#### Service Account anlegen

Das `service-account-create`-Skript verwaltet das Erstellen eines neuen Service Accounts. Es wird immer ausgeführt, wenn ein neuer Konsument installiert wird. Während des Aufrufs erhält das `service-account-create`-Skript jedenfalls den Namen des Konsumenten-Dogus als erstes Argument. Ein zweites Argument zur Feineinstellung des Service Accounts ist optional und hängt von der Unterstützung des Produzenten-Dogus ab.

Ein `service-account-create`-Skript besteht aus drei Schritten:

> 1. Erstellen der Account Informationen

Reviewer: ```suggestion
1. Erstellen der Account-Informationen
```

Reviewer 2: Knapp darunter wird es einfach zusammen geschrieben

Author: done

2. Anlegen eines Accounts für unsere Software mit den Informationen

> 3. Schreibe die Accountinformationen verschlüsselt in den Dogu-spezifischen Konfigurationsbereich des Verbrauchers.

Reviewer: ```suggestion
3. Verschlüsseltes Schreiben der Accountinformationen in den Dogu-spezifischen Konfigurationsbereich des Verbrauchers
```

Author: Done. I rephrased the whole sentence  to emphasize the actual task at hand

Ein Beispiel für das `service-account-create`-Skript könnte folgendermaßen aussehen:

`create-sa.sh`

```bash
#!/bin/bash
set -o errexit
set -o nounset
set -o pipefail
doguName="${1}"
{
  #1) Erstellen der Zugangsdaten
  USER="${doguName}-$(doguctl random)"
  PASSWORD=$(doguctl random)

  #2) Legt einen Account für unsere Software mit den Informationen an
  # ... Implementiere die Dogu-spezifische Logik hier
} >/dev/null 2>&1 # unterdrücke sonstige Ausgaben auf stdout, da nur Zugangsdaten oder Fehler erwartet werden

  #3) Gib Zugangsdaten zurück. Der Client schreibt die Zugangsdaten verschlüsselt in den Dogu-spezifischen Konfigurationsbereich des Verbrauchers.
echo "username: ${USER}"
echo "password: ${PASSWORD}"
```

In Schritt 1 werden Zugangsdaten zufällig generiert. Der Name des Konsumenten-Dogus kann in die Zugangsdaten eingearbeitet werden, allerdings ist dies keine Pflicht.

In Schritt 2 muss ein Account mit den generierten Informationen in der Dogu-Software erstellt werden. 

> In Schritt 3 werden die Zugangsdaten mit dem Format `echo "<registrykey>: <registryvalue>"` ausgegeben. `registrykey` und `registryvalue` müssen durch einen Doppelpunkt und einem Leerzeichen getrennt sein. Die gesamte Zeile muss durch ein Newline beendet werden.

Reviewer: ```suggestion
In Schritt 3 werden die Zugangsdaten mit dem Format `echo "<registrykey>: <registryvalue>"` ausgegeben. `registrykey` und `registryvalue` müssen durch einen Doppelpunkt und einem Leerzeichen getrennt sein. Die gesamte Zeile muss durch einen Zeilenumbruch beendet werden.
```

Author: done

Diese Zugangsdaten werden vom verarbeitenden Client automatisch eingelesen und in den Dogu-spezifischen Konfigurationsbereich des Verbrauchers verschlüsselt abgelegt.

> Nach dem obigen Beispiel würde dem Konsumenten-Dogu zwei Registry-Einträge einrichten, die das jeweilige Zugangsdatum als Wert enthalten:  

Reviewer: ```suggestion
Nach dem obigen Beispiel würde das Konsumenten-Dogu zwei Registry-Einträge einrichten, die das jeweilige Zugangsdatum als Wert enthalten:  
```

Reviewer 2: Better for the english translation as this could easily be overseen as 'access date'.
```suggestion
Nach dem obigen Beispiel würde das Konsumenten-Dogu je einen Registry-Eintrag für Username und Passwort erzeugen:  
```

Author: Done. I chose a different word

Reviewer: argh, I changed it again :D
See 24488a4c776273cc16b6228eae2dfdc714dcd198

- `/config/<konsument>/sa-<produzent>/username`

> - `/config/<konsument>/sa-<produzent>/passwort` 

Reviewer 2: A german etcd key?
```suggestion
- `/config/<konsument>/sa-<produzent>/password` 
```

Author: good catch

Das Verbaucher-Dogu kann diese nun z. B. durch `doguctl config -e sa-<produzentdogu>/username` auslesen und entschlüsseln.

#### Service Account löschen

Das `service-account-remove`-Skript löscht einen bestehenden Service Account eines Konsumenten-Dogus.

Es wird immer ausgeführt, wenn das Konsumenten-Dogu deinstalliert wird. Während des Aufrufs erhält das `service-account-remove`-Skript jedenfalls den Namen des Konsumenten-Dogus als erstes Argument. Da hier keine Ausgaben des Skriptes verarbeitet werden, können auch Fehlermeldungen auf `stdout` ausgegeben werden.

Ein `service-account-remove`-Skript besteht aus zwei Schritten:

> 1. Identifier von dem Konsumenten identifizieren 

Reviewer: Maybe find a better word than "Identifier"

Reviewer 2: ```suggestion
1. Service-Account-Konsument identifizieren 
```

Author: done

2. Löschen der Daten und des Service Accounts des Konsumenten

Ein Beispiel für das `service-account-remove`-Skript könnte folgendermaßen aussehen:

`remove-sa.sh`

```bash
#!/bin/bash
set -o errexit
set -o nounset
set -o pipefail

> #1) Identifier von dem Konsumenten identifizieren

Reviewer: s.o.

Author: done

SERVICE="${1}"
if [ X"${SERVICE}" = X"" ]; then
  echo "usage remove-sa.sh servicename"
  exit 1
fi

#2) Löschen des Service Accounts des Verbrauchers
# ... Implementiere die Dogu-spezifische Logik hier
```

> In Schritt 1 wird der Name des Konsumenten-Dogus ausgelesen (z. B. "redmine"), i. d. R. um den Service Account des Dogus zu identifizieren.

Reviewer: ```suggestion
In Schritt 1 wird der Name des Konsumenten-Dogus ausgelesen (z. B. "redmine"), in der Regel um den Service Account des Dogus zu identifizieren.
```

Author: I think the whole phrase can be removed.

In Schritt 2 wird der Account aus der Datenbasis des Produzenten inkl. Zugangsdaten gelöscht.

### Service Accounts konsumieren

> Es ist sehr einfach, bei einem Produzenten-Dogu einen Service Account anzufragen, da die Hauptarbeit beim Client und beim Produzenten liegt. Ein Konsumenten-Dogu muss lediglich in seiner `dogu.json` das gewünschte Produzenten-Dogu als [ServiceAccount]() nennen:

Reviewer: Linkziel fehlt

Author: done

```json
{
  ...,
  "ServiceAccounts": [
    {
      "Type": "produzentendogu",
      "Kind": "dogu"
    }
  ],
  ...
}
```

> Bei einem Service Account handelt es sich um eine besondere Form der Abhängigkeit. Trotzdem ist es sinnvoll, das Produzenten-Dogu in seiner [Dependency-Liste]() zu führen. So wird in den unterschiedlichen Phasen einer Dogu-Installation sichergestellt, dass das Produzenten-Dogu wirklich zur Benutzung bereitsteht:

Reviewer: Linkziel fehlt

Author: done

> Bei einem Service Account handelt es sich um eine besondere Form der Abhängigkeit. Trotzdem ist es sinnvoll, das Produzenten-Dogu in seiner [Dependency-Liste]() zu führen. So wird in den unterschiedlichen Phasen einer Dogu-Installation sichergestellt, dass das Produzenten-Dogu wirklich zur Benutzung bereitsteht:

Reviewer 2: ```suggestion
Bei einem Service Account handelt es sich um eine besondere Form der Abhängigkeit. Deshalb ist es sinnvoll, das Produzenten-Dogu in seiner [Dependency-Liste]() zu führen. So wird in den unterschiedlichen Phasen einer Dogu-Installation sichergestellt, dass das Produzenten-Dogu wirklich zur Benutzung bereitsteht:
```

Author: Done

```json
{
  ...,
  "Dependencies": [
    {
      "name": "produzentendogu",
      "type": "dogu"
    }
  ],
  "ServiceAccounts": [
    {
      "Type": "produzentendogu",
      "Kind": "dogu"
    }
  ],
  ...
}
```

## Dogu-Upgrades

> Das Upgrade eines Dogus kann mit `cesapp upgrade dogu <doguname>` durchgeführt werden. Dabei verwaltet die cesapp alle wichtigen Schritte, um das Dogu erfolgreich zu upgraden.

Author: I would like to abstract here from `cesapp`, since there is a K8s-centric CES solution on the horizon. We use to speak from "clients" that install, upgrade or delete dogus.

Reviewer: resolved by fc6c870ef6216fda1079f061f8049a204eaf2c6e

> Der Upgrade-Prozess enthält u.A. folgende Schritte:

Author: Abbreviations that contain unabbreviated a space keep the space when abbreviated.

```suggestion
Der Upgrade-Prozess enthält u. A. folgende Schritte:
```

- Anzeigen von Upgrade Notifications, falls vorhanden
- Download des neuen Dogu-Images
- Sicherstellen, dass das Dogu healthy ist
- Sicherstellen, dass alle abhängigen Dogus verfügbar sind

> - Ausführen des Pre-Upgrade-Skripts im alten Dogu, falls vorhanden

Reviewer: The following paragraphs explain this in depth (see line 986). This here is only a short summary that I wouldn't clutter with links.

- Ersetzen des alten Dogus durch das neue
- Starten des neuen Dogus
- Ausführen des Post-Upgrade-Skripts im neuen Dogu, falls vorhanden

Falls es beim Dogu-Upgrade notwendig ist, vor- oder nachbereitende Schritte durchzuführen, können optionale Pre- und Post-Upgrade-Skripte erstellt werden.

> Als Beispiel kann man eine Software betrachten, die in einer neueren Version eine andere Struktur der Datenbank benötigt.

Reviewer 2: ```suggestion
Als Beispiel kann man eine Software betrachten, die nach einem Update ein anderes Datenbankschema benötigt.
```

In einem solchen Fall ist es möglich, Upgrade-Skripte zu schreiben, um eine Migration der Datenbank nach einem Dogu-Upgrade zu starten.

> Außerdem kann man über wichtige Änderungen beim Upgrade eines Dogus durch ein Upgrade-Notification-Skript hinweisen lassen.

Author: I suggest to use either "Außerdem kann man sich ... hinweisen lassen" or "Außerdem kann man ... hinweisen"

Reviewer: Als Entwickler willst du ja durch das Skript nicht DICH auf etwas hinweisen lassen, sondern die Anwender/Administratoren.

> Alle im Folgenden beschriebenen Upgrade-Skripte sollten (neben bspw. dem `startup.sh`-Skript) im `resources`-Ordner abgelegt werden.

Reviewer: It's best practice and keeps a good structure. Yes, you can put your scripts anywhere, but why would you, if it's working so well in all of the Dogus? I think "sollten" does reflect this.

> Alle im Folgenden beschriebenen Upgrade-Skripte sollten (neben bspw. dem `startup.sh`-Skript) im `resources`-Ordner abgelegt werden.

Author: ok

### Ermittlung der Upgrade-Pfade

Bei der Entwicklung der Dogus ist dafür zu sorgen, dass Upgrades eines Dogus auch über mehrere Versionen hinweg funktionieren. Dafür müssen die Upgrade-Skripte so gestaltet sein, dass sie den Upgradepfad ermitteln und ggf. nötige Migrationsschritte automatisch ausführen.

Hierfür können in allen Upgrade-Skripten die vorherige (`FROM`) und neue (`TO`) Dogu-Version ausgelesen werden:

```bash
FROM_VERSION="${1}"
TO_VERSION="${2}"
```

> ### pre-upgrade.sh - Führt alle Aktionen vor dem Upgrade des Dogus durch

Reviewer: done

Das Pre-Upgrade-Skript kann als [Exposed Command](../core/compendium_de.md#exposedcommands) in der dogu.json eines Dogus definiert werden:

```json

>   "ExposedCommands": [
>     {
>       "Name": "pre-upgrade",
>       "Command": "/pre-upgrade.sh"
>     }
>   ]

Author: I suggest to hint the dogu developer that there is more to the `dogu.json` than this snippet, like so:

```suggestion
{
   ...,
  "ExposedCommands": [
    {
      "Name": "pre-upgrade",
      "Command": "/pre-upgrade.sh"
    }
  ],
  ...
  }
```

```

Dieses Skript wird vor dem eigentlichen Upgrade des Dogus im alten Dogu-Container ausgeführt.

> ### post-upgrade.sh - Führt alle Aktionen nach dem Upgrade des Dogus durch

Reviewer: done

Das Post-Upgrade-Skript kann als [Exposed Command](../core/compendium_de.md#exposedcommands) in der dogu.json eines Dogus definiert werden:

```json

>   "ExposedCommands": [
>     {
>       "Name": "post-upgrade",
>       "Command": "/post-upgrade.sh"
>     }
>   ]

Author: I suggest to hint the dogu developer that there is more to the `dogu.json` than this snippet, like so:
```suggestion
{
  ...,
  "ExposedCommands": [
    {
      "Name": "post-upgrade",
      "Command": "/post-upgrade.sh"
    }
  ],
  ...,
}
```

```

> Dieses Skript wird nach dem Upgrade des Dogus im neuen Dogu-Container gestartet, sobald dieser hochfährt. Es laufen also im Normalfall sowohl das `startup.sh`- als auch das `post-upgrade.sh`-Skript gleichzeitig los. Falls es notwendig sein sollte, dass das `post-upgrade.sh`-Skript noch vor dem `startup.sh`-Skript ausgeführt wird, muss das `startup.sh`-Skript mit einem Wartemechanismus versehen werden. Dies kann bspw. durch das Warten auf einen etcd-Key geschehen, den das `post-upgrade.sh`-Skript setzt, sobald es komplett durchgelaufen ist.

Author: etcd vs Cloudogu EcoSystem registry, ExposedCommand identifier vs upgrade script name

- We should stop using etcd and registry synonymous because the KV store could be exchanged.
- like above the dogu developer may choose a different file name for a post-upgrade script. Our docs should reflect this.
```suggestion
Dieses Skript wird nach dem Upgrade des Dogus im neuen Dogu-Container gestartet, sobald dieser hochfährt. Es laufen also im Normalfall sowohl das `startup.sh`- als auch das `post-upgrade`-Skript gleichzeitig los. Falls es notwendig sein sollte, dass das `post-upgrade`-Skript noch vor dem `startup.sh`-Skript ausgeführt wird, muss das `startup.sh`-Skript mit einem Wartemechanismus versehen werden. Dies kann bspw. durch das Warten auf einen Registry-Key geschehen, den das `post-upgrade`-Skript setzt, sobald es komplett durchgelaufen ist.
```

### upgrade-notification.sh - Zeigt eine Benachrichtigung vor der Upgradebestätigung eines Dogus

> Das Upgrade-Notification-Skript kann als [Exposed Command](../core/compendium_de.md#exposedcommands) in der dogu.json eines Dogus definiert werden:

Author: File references with special format for easier recognition?
```suggestion
Das Upgrade-Notification-Skript kann als [Exposed Command](../core/compendium_de.md#exposedcommands) in der `dogu.json` eines Dogus definiert werden:
```

```json

>   "ExposedCommands": [
>     {
>       "Name": "upgrade-notification",
>       "Command": "/upgrade-notification.sh"
>     }
>   ]

Author: I suggest to hint the dogu developer that there is more to the dogu.json than this snippet, like so:
```suggestion
{
  ...,
  "ExposedCommands": [
    {
      "Name": "upgrade-notification",
      "Command": "/upgrade-notification.sh"
    }
  ]
  ...,
}
```

```

Dieses Skript wird vor dem Upgrade-Vorgang ausgeführt und sollte ausschließlich Informationen (per `echo`) ausgeben, die für den Administrator vor dem Upgrade relevant sein können. Beispielsweise kann hier auf Breaking Changes hingewiesen werden oder darauf, dass man vor dem Upgrade ein Backup anlegen sollte.

## Typische Dogu-Features

> Dieses Kapitel beschreibt Funktionen, die Dogus tiefer in das Cloudogu EcoSystem integrieren und sie einheitlich administrierbar machen.

Author: What a lovely introduction :blue_heart:

### Memory-/Swap-Limit

> Mit Memory- und Swap-Limits kann man den Speicherverbrauch (Arbeitsspeicher und Swap) von Dogus einschränken.

Author: Maybe describe once the term "swap"?

```suggestion
Mit Memory- und Swap-Limits kann man den Speicherverbrauch (Arbeitsspeicher und Auslagerungsspeicher) von Dogus einschränken.
```

Falls ein Dogu seine Speicherlimitierung überschreitet, wird der größte Prozess im Container beendet.
Dies ist normalerweise der Hauptprozess des Dogus und führt dazu, dass der Container neu gestartet wird.

Wird kein Wert bei der Memory-Limitierung gesetzt, findet diese auch nicht statt.
Bei der Swap-Limitierung is `0b` der Standardwert und stellt somit keinen Swap zur Verfügung.

> Um eine Limitierung vornehmen zu können, muss die dogu.json des Dogus folgende Einträge enthalten:

Author: File references with special format for easier recognition?
```suggestion
Um eine Limitierung vornehmen zu können, muss die `dogu.json` des Dogus folgende Einträge enthalten:
```

```json

> {
>     "Name": "container_config/memory_limit",
>     "Description":"Limits the container's memory usage. Use a positive integer value followed by one of these units [b,k,m,g] (byte, kibibyte, mebibyte, gibibyte).",
>     "Optional": true,
>     "Validation": {
>     "Type": "BINARY_MEASUREMENT"
>     }
> },
> {
>     "Name": "container_config/swap_limit",
>     "Description":"Limits the container's swap memory usage. Use zero or a positive integer value followed by one of these units [b,k,m,g] (byte, kibibyte, mebibyte, gibibyte). 0 will disable swapping.",
>     "Optional": true,
>     "Validation": {
>     "Type": "BINARY_MEASUREMENT"
>     }
> }

Author: Hinting the full json format would have prevented weird indentation. I suggest to add outer braces like so
```suggestion
{
...,
  {
    "Name": "container_config/memory_limit",
    "Description":"Limits the container's memory usage. Use a positive integer value followed by one of these units [b,k,m,g] (byte, kibibyte, mebibyte, gibibyte).",
    "Optional": true,
    "Validation": {
    "Type": "BINARY_MEASUREMENT"
    }
},
{
    "Name": "container_config/swap_limit",
    "Description":"Limits the container's swap memory usage. Use zero or a positive integer value followed by one of these units [b,k,m,g] (byte, kibibyte, mebibyte, gibibyte). 0 will disable swapping.",
    "Optional": true,
    "Validation": {
    "Type": "BINARY_MEASUREMENT"
    }
  }
  ...,
}
```

```

> Hiermit lassen sich die etcd-Einträge `container_config/memory_limit` und `container_config/swap_limit` in der jeweiligen Dogu-Konfiguration erstellen.

Author: etcd vs Cloudogu EcoSystem registry
```suggestion
Hiermit lassen sich die Registry-Einträge `container_config/memory_limit` und `container_config/swap_limit` in der jeweiligen Dogu-Konfiguration erstellen.
```

Die konfigurierbaren Werte für die Schlüssel sind jeweils eine Zeichenkette der Form `<Zahlenwert><Einheit>` und beschreiben die maximal vom Dogu nutzbare Menge an Speicher.
Zu beachten ist hier, dass zwischen dem numerischen Wert und der Einheit kein Leerzeichen stehen darf.
Verfügbare Einheiten sind `b`, `k`, `m` und `g` (für byte, kibibyte, mebibyte und gibibyte).

Das Setzen der Werte kann über folgende Wege erfolgen:
- `doguctl config container_config/memory_limit 1g`
- `cesapp edit-config <doguname>` (nur vom Host aus)
- `etcdctl set /config/<doguname>/container_config/memory_limit "1g"` (nur vom Host aus)

Um die Limitierungen zu übernehmen, muss das Dogu neu erstellt (`cesapp recreate <doguname>`) und anschließend neu gestartet (`cesapp start <doguname>`) werden.

> Ein Sonderfall stellt die Limitierung eines Java-Prozesses dar. Enthält ein Dogu einen Java-Prozess, können folgende zusätzliche Einträge in die dogu.json eingebaut werden:

Author: File references with special format for easier recognition?
```suggestion
Ein Sonderfall stellt die Limitierung eines Java-Prozesses dar. Enthält ein Dogu einen Java-Prozess, können folgende zusätzliche Einträge in die `dogu.json `eingebaut werden:
```

```json

> {
>   "Name": "container_config/java_max_ram_percentage",
>   "Description": "Limits the heap stack size of the Java process to the configured percentage of the available physical memory when the container has more than approx. 250 MB of memory available. Is only considered when a memory_limit is set. Use a valid float value with decimals between 0 and 100 (f. ex. 55.0 for 55%). Default value: 25%",
>   "Optional": true,
>   "Default": "25.0",
>   "Validation": {
>     "Type": "FLOAT_PERCENTAGE_HUNDRED"
>   }
> },
> {
>   "Name": "container_config/java_min_ram_percentage",
>   "Description": "Limits the heap stack size of the Java process to the configured percentage of the available physical memory when the container has less than approx. 250 MB of memory available. Is only considered when a memory_limit is set. Use a valid float value with decimals between 0 and 100 (f. ex. 55.0 for 55%). Default value: 50%",
>   "Optional": true,
>   "Default": "50.0",
>   "Validation": {
>     "Type": "FLOAT_PERCENTAGE_HUNDRED"
>   }
> }

Author: Hinting the full json format would have prevented weird indentation. I suggest to add outer braces like so

```suggestion
{
  ...,
  "Configuration": [
    {
      "Name": "container_config/java_max_ram_percentage",
      "Description": "Limits the heap stack size of the Java process to the configured percentage of the available physical memory when the container has more than approx. 250 MB of memory available. Is only considered when a memory_limit is set. Use a valid float value with decimals between 0 and 100 (f. ex. 55.0 for 55%). Default value: 25%",
      "Optional": true,
      "Default": "25.0",
      "Validation": {
        "Type": "FLOAT_PERCENTAGE_HUNDRED"
      }
    },
    {
      "Name": "container_config/java_min_ram_percentage",
      "Description": "Limits the heap stack size of the Java process to the configured percentage of the available physical memory when the container has less than approx. 250 MB of memory available. Is only considered when a memory_limit is set. Use a valid float value with decimals between 0 and 100 (f. ex. 55.0 for 55%). Default value: 50%",
      "Optional": true,
      "Default": "50.0",
      "Validation": {
        "Type": "FLOAT_PERCENTAGE_HUNDRED"
      }
    }
  ],
  ...,
}
```

```

Die damit konfigurierbaren Werte müssen in den Start-Skripten des Dogus dem entsprechenden Java-Prozess als Parameter mitgegeben werden. Eine Referenzimplementierung findet sich im [Nexus-Dogu](https://github.com/cloudogu/nexus/blob/77bdcfdbe0787c85d2d9b168dc38ff04b225706d/resources/util.sh#L52).

### Backup & Restore-Fähigkeit

Ein Feature des Cloudogu EcoSystems ist, dass alle Dogus so ausgelegt sind, dass sie sich über ein zentrales Backupsystem sichern und wiederherstellen lassen. Damit das funktioniert, müssen alle Dogus ihre veränderlichen Daten in Volumes auslagern. Dieses Vorgehen wird bereits im [Compendium](../core/compendium_de.md#volumes) beschrieben.

> Bei der Entwicklung eines Dogus ist darauf zu achten, dass das Dogu nach einem erfolgreichen Backup- und Restore-Vorgang weiterhin normal funktioniert und alle Daten und Funktionen vorhanden sind. Dazu müssen zuerst alle Volumes in der dogu.json, die Produktivdaten enthalten, mit dem [NeedsBackup](../core/compendium_de.md#needsbackup)-Flag gekennzeichnet werden. Danach sollte das Dogu gebaut und mit Testdaten gefüllt werden.

Author: ```suggestion
Bei der Entwicklung eines Dogus ist darauf zu achten, dass das Dogu nach einem erfolgreichen Backup- und Restore-Vorgang weiterhin normal funktioniert und alle Daten und Funktionen vorhanden sind. Dazu müssen zuerst alle Volumes in der `dogu.json`, die Produktivdaten enthalten, mit dem [NeedsBackup](../core/compendium_de.md#needsbackup)-Flag gekennzeichnet werden. Danach sollte das Dogu gebaut und mit Testdaten gefüllt werden.
```

> Anschließend führt man ein Backup des Systems und danach einen Restore durch. Dazu können die Befehle [cesapp backup](https://docs.cloudogu.com/de/docs/system-components/cesapp/operations/backup/) und [cesapp restore](https://docs.cloudogu.com/de/docs/system-components/cesapp/operations/restore/) genutzt werden.

Author: This reads funnily. I suppose developers know the sequence of back-up and restore. Then the first sentence could be simplified:

```suggestion
Anschließend führt man ein Backup und ein Restore des Systems durch. Dazu können die Befehle [cesapp backup](https://docs.cloudogu.com/de/docs/system-components/cesapp/operations/backup/) und [cesapp restore](https://docs.cloudogu.com/de/docs/system-components/cesapp/operations/restore/) genutzt werden.
```

Sind alle Dogus wieder hochgefahren, testet man, ob das eigene Dogu normal läuft und alle Testdaten weiterhin vorhanden sind.

### Änderbarkeit der Admin-Gruppe

> Bei der Installation des CES wird eine globale Admin-Gruppe eingestellt. Alle Mitglieder dieser Gruppe sollen in allen Dogus Admin-Rechte bekommen. Dies muss bei der Entwicklung des Dogus sichergestellt werden. Der Name der global definierten Admin-Gruppe lässt sich aus dem Dogu heraus mit `doguctl` abfragen: `doguctl config --global 'admin_group'`

Author: CES is quite informal. We should call our product by its name.
```suggestion
Bei der Installation des Cloudogu EcoSystems wird eine globale Admin-Gruppe eingestellt. Alle Mitglieder dieser Gruppe sollen in allen Dogus Admin-Rechte bekommen. Dies muss bei der Entwicklung des Dogus sichergestellt werden. Der Name der global definierten Admin-Gruppe lässt sich aus dem Dogu heraus mit `doguctl` abfragen: `doguctl config --global 'admin_group'`
```

### Änderbarkeit der FQDN

Die FQDN des CES-Systems ist global im etcd gespeichert und kann von den Dogus ausgelesen (`doguctl config --global fqdn`) und in die Dogu-Konfiguration integriert werden, falls nötig (siehe auch [Erstellung einer exemplarischen Anwendung](../core/basics_de.md#5-erstellung-einer-exemplarischen-anwendung). Zusätzlich muss bei der Dogu-Entwicklung darauf geachtet werden, dass die FQDN änderbar ist. Das Dogu sollte also in der Lage sein (ggf. nach einem Neustart), eine neue FQDN auszulesen und seine Konfiguration an diese neue FQDN anzupassen.

### Logging-Verhalten steuern

Dieser Abschnitt zeigt, wie eine einheitliche, abstrahierte Einstellmöglichkeit geschaffen wird, mit der man das Log-Verhalten in allen Dogus auf die gleiche Art und Weise einstellen kann. Dazu werden vier verschiedene über die etcd-Registry setzbare Log-Level genutzt: `ERROR`, `WARN`, `INFO` und `DEBUG`.

> Mit diesen abstrakten Log-Leveln soll der Administrator unterstützt werden. Er muss nicht wissen, wie ein Dogu-spezifisches Log-Level heißt, sondern kann stattdessen auf vier Grundwerte zurückgreifen und diese gefahrlos in allen Dogus wiederverwenden.

Author: A fantastic explanation! :blue_heart:

> Als Entwickler eines Dogus müssen wir darauf achten, dass diese vier Log-Level sinnvoll auf die Log-Level der Software im Dogu gemappt werden, welche ggf. andere Namen haben können (bspw. `TRACE` oder `FATAL`).

Author: The rest of the documentation is not formulated from a we-perspective.
```suggestion
Dogu-Entwickler müssen darauf achten, dass diese vier Log-Level sinnvoll auf die Log-Level der Software im Dogu gemappt werden, welche ggf. andere Namen haben können (bspw. `TRACE` oder `FATAL`).
```

- eine relevante Entscheidung für eine Konfiguration wurde gefällt
- Meldungen mit `TRACE`-Log-Level

#### Mapping je Dogu

Diese vier abstrakten Log-Levels haben hin und wieder keine direkte Beziehung zu konkreten Log-Levels in den Tools oder Plugins im Dogu und müssen in der Dogu-Entwicklung manuell abgebildet werden. Der Grund dafür liegt darin, dass sich sowohl von den Tools als auch damit einhergehenden Plugins und Frameworks die Log-Levels zu stark in Quantität und Qualität unterscheiden.

Die Abbildung eines abstrakten Log-Levels liegt damit in der Hand der Dogu-Entwicklung. Hier kann man beispielsweise entscheiden, dass man das `TRACE`-Log-Level aus der Software auf das `DEBUG`-Log-Level des Dogus mappen möchte.

Diese Abbildung von abstrakten zu konkreten Log-Levels birgt den Vorteil der Flexibilität, z. B. wenn mehrere unterschiedliche Konfigurationsdateien mit unterschiedlichen Log-Leveln gefüllt werden müssen.

#### Wertebereich und Default

Zum Einstellen des Log-Levels wird je Dogu ein optionaler Key `/config/${dogu}/logging/root` eingeführt, der einen der folgenden abstrakten Log-Level-Werte annehmen kann.

- `ERROR`
- `WARN` (Defaultwert)
- `INFO`
- `DEBUG`

> Wenn kein Key `/config/${dogu}/logging/root` existiert, dann wird automatisch `WARN` verwendet.

Author: Maybe it is worth a shot to mention who uses the default value if the `dogu.json` is configured properly?
```suggestion
Wenn kein Key `/config/${dogu}/logging/root` existiert, dann verwendet `doguctl` automatisch den Wert `WARN`, wenn dieser ordnungsgemäß [als Default-Wert](#validierung-und-default-werte) konfiguriert wurde.
```
